NetKet/UnitaryTimeEvolution/TimeStepper/controlled_time_steppers.hpp

```diff
@@ -0,0 +1,266 @@
+    double last_norm_;
+
+    static constexpr double safety_factor = 0.95;
+    static constexpr double err_exponent = -1. / 5.;
+
+public:
+    void Propagate(OdeSystemFunction<State> ode_system,
+                   State &x,
+                   Time t, Time dt) final override
+    {
+        assert(dt > .0);
+        Time tmax = t + dt;
+
+        last_x_.resize(x.size());
+        last_x_ = x;
+        last_norm_ = x.norm();
+
+        double current_t = t;
+        if(current_dt_ == .0)
+        {
+            current_dt_ = dt / 10.; // TODO: better way to guess initial step size?
+        }
+        while(current_t < tmax)
+        {
+            Time next_dt = (current_t + current_dt_ < tmax) ? current_dt_ : tmax - current_t;
+
+            double scaled_error = PerformSingleStep(ode_system, x, current_t, next_dt);
+
+            if(scaled_error <= 1.0) // error is within bounds
+            {
+                current_t += next_dt;
+                last_x_ = x;
+                last_norm_ = x.norm();
+            }
+            else // error too large, try again with new step size
+            {
+                x = last_x_;
+                Reset();
+            }
+
+            // update time step
+            double dt_factor = safety_factor * std::pow(scaled_error, err_exponent);
+            dt_factor = clamp(dt_factor, 0.01, 10.);
+            current_dt_ *= dt_factor;
+        };
+    }
+
+protected:
+    ControlledStepperBase(double atol, double rtol)
+            : atol_(atol), rtol_(rtol), current_dt_(0.), last_norm_(0.)
+    {
+    }
+
+    /**
+     * Perform a single step from t to t + dt.
+     * @param ode_system The ODE system function.
+     * @param x The current state, which is modified in place.
+     * @param t The current time.
+     * @param dt The time step.
+     * @return The scaled error for the performed step. This should be calculated
+     *      by calling LocalError.
+     */
+    virtual double PerformSingleStep(OdeSystemFunction<State> ode_system,
+                                     State &x,
+                                     Time t, Time dt) = 0;
+
+    /**
+     * Reset the internal state of the stepper, in case the next time step
+     * does not start at the end of the previous time step.
+     */
+    virtual void Reset() = 0;
+
+    /**
+     * Compute the relative local error from the error estimate delta, taking into account
+     * the atol and rtol values of the stepper.
+     * @param delta A local error estimate, usually computed as the difference between
+     *      the resulting state for two different-order methods.
+     * @param norm The norm of the current state, used for rescaling rtol.
+     * @return The relative local error of the solution. This will be a
+     *      non-negative number that is <= 1 iff the error is within the
+     *      bound given by atol and rtol.
+     */
+    double LocalError(const State& delta, double norm) const
+    {
+        assert(norm >= .0);
```

Comment:
It might be a good idea to add something like `assert(delta.size() > 0)` here, because otherwise you might end up dividing by zero.